File: solution/1552. Magnetic Force Between Two Balls/1552.cpp
class Solution {
public:
    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(), position.end());
        
        int left = 0, right = position.back() - position[0], mid;
        while (left < right) {
            mid = right - (right - left) / 2;
            if (checkOK(position, m, mid))
                left = mid;
            else
                right = mid - 1;
        }
        return left;
    }
    
    bool checkOK(vector<int>& P, int m, int d) {
        if (m-- == 1) return true;
        int pre = P[0];
        for (int i = 1; i < P.size(); ++i) {
            if (P[i] - pre >= d) {
                if (--m == 0)
                    return true;
                pre = P[i];
            }
        }
        return false;
    }
};

File: solution/1552. Magnetic Force Between Two Balls/1552.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        n = len(position)
        position.sort()
        
        def check_ok(d):
            cnt = 1
            pre = position[0]
            
            for i in range(1, n):
                if position[i] - pre >= d:
                    cnt += 1
                    if cnt >= m: return True
                    pre = position[i]
            return False
        
        l, r = 0, position[-1] - position[0]
        while l < r:
            mid = r - (r - l) // 2
            if check_ok(mid):
                l = mid
            else:
                r = mid - 1
        return l
    
